README.md
# haskell-tic-tac-roll
##### By Daniel Hao Huynh
Use `stack run` in order to start the application
> gamemode is coded in application instead of pipes

#### Testing
> Used Hspec, Doctest, HUnit, QuickTest


ChangeLog.md
# Changelog for haskell-tic-tac-roll

## Unreleased changes


test/Spec.hs
import Test.Hspec ( hspec, context, describe, it, shouldBe )
import Control.Exception(evaluate)
import Test.QuickCheck (quickCheck)
import Test.DocTest (doctest)
import Test.HUnit (runTestTT, Test (TestCase, TestLabel, TestList), assertEqual)
import System.Random (mkStdGen, Random (randomR), StdGen)
import Data.Time.Clock (getCurrentTime, utctDayTime)
import GameHandler (
    countOcc,
    convertInn,
    printBoard,
    swapEarlyMarks,
    createBoard,
    readInn,
    readUnit,
    updateBoard,
    roll,
    checkBoard,
    gameLoopErrHandlr,
    getPlayerTurn,
    genRandNum,
    checkDraw
  )
import ComputerOperatedPlayer (
    readStr,
    readCOP,
    copHandleRoll,
    copGetLegalSpace,
    copErr
  )

prop_createBoard:: [Char] -> Bool
prop_createBoard board = createBoard == "         " 
 

main :: IO ()
main = do
  doctest ["-isrc", "src/ComputerOperatedPlayer.hs", "src/GameHandler"] -- DocTest
  
  quickCheck prop_createBoard
  _ <- runTestTT unitTests -- UnitTest
  
  hspec $ do -- Hspec
    describe "countOcc" $ do
      it "counts occurrances of a char from string" $ do
        countOcc 'J' "Jason, Jason Bourne" `shouldBe` 2
        countOcc 'B' "BOB DYLAN IS OP" `shouldBe` 2

      context "if no char matches" $
        it "returns 0" $
          countOcc '!' "" `shouldBe` 0

    describe "createBoard" $ do
      it "creates a new board" $ do
        createBoard `shouldBe` "         "

    describe "updateBoard" $ do
      it "updates board based on index data and char" $ do
        updateBoard createBoard (0,'X') `shouldBe` "X        "
        updateBoard "O X OXX  " (0,'X') `shouldBe` "X X OXX  "

    describe "printBoard" $ do
      it "converts raw board data to printable kawaii string" $ do
        printBoard createBoard `shouldBe` " \t   1  2  3 \n#\t1 [ ][ ][ ]\n#\t2 [ ][ ][ ]\n#\t3 [ ][ ][ ]"
        printBoard "O X OXX  " `shouldBe` " \t   1  2  3 \n#\t1 [O][ ][X]\n#\t2 [ ][O][X]\n#\t3 [X][ ][ ]"

    describe "swapEarlyMarks" $ do
      it "swaps the first and third mark on board" $ do
        swapEarlyMarks createBoard `shouldBe` "         "
        swapEarlyMarks "O X OXX  " `shouldBe` "X O OXX  "

    describe "roll" $ do
      it "swaps the first and third mark on board then rolls board to a direction" $ do
        roll createBoard "left" `shouldBe` createBoard
        roll "O X OXX  " "right" `shouldBe` "X X O  XO"
        roll "O X OXX  " "left" `shouldBe`  "OX  O X X"

    describe "checkDraw" $ do
      it "checks board for draw" $ do
        checkDraw createBoard 0 `shouldBe` False
        checkDraw "OXOXOXXOX" 0 `shouldBe` True
        checkDraw "O X OXX  " 0 `shouldBe`  False

    describe "checkBoard" $ do
      it "checks board for win condition" $ do
        checkBoard createBoard  `shouldBe` (False,' ',"")
        checkBoard "OXOXOXXOX"  `shouldBe` (False,' ',"")
        checkBoard "O X OXX X"  `shouldBe` (True,'X',"O | O|X |")

    describe "readUnit" $ do
      it "reads string and checks for err " $ do
        readUnit "1"  `shouldBe` 1
        readUnit "1 12311"  `shouldBe` 1
        readUnit "1 12311a"  `shouldBe` 1

      context "if err is encountered" $ do
        it "returns 13" $ do
          readUnit ""  `shouldBe` 13 --err
          readUnit "1a"  `shouldBe` 13 --err

    describe "getPlayerTurn" $ do
      it "gets current turn and returns current- and next turn" $ do
        getPlayerTurn 'X'  `shouldBe` ('X','O')
        getPlayerTurn 'O'  `shouldBe` ('O','X')

    describe "gameLoopErrHandlr" $ do
      it "Gets raw index data for board, and checks for err" $ do
        gameLoopErrHandlr 11 `shouldBe` True
        gameLoopErrHandlr 12 `shouldBe` True
        gameLoopErrHandlr 13 `shouldBe` True
      context "if no err is encountered" $
        it "returns False" $
          gameLoopErrHandlr 1  `shouldBe` False

    describe "convertInn" $ do
      it "Converts input to data nessescary for board. and checks for err" $ do
        convertInn ["1","2"] `shouldBe` (1,2,"2")
      context "if roll is encountered" $
        it "returns with roll" $
          convertInn ["1","2", "right"]  `shouldBe` (1,2,"right")
      context "if incorrect format encountered" $
        it "returns err code" $
          convertInn ["a","a"]  `shouldBe` (13,13,"a")

    describe "genRandNum" $ do
      it "gets a random num based on min max and seed" $ do
        genRandNum 1 2 1  `shouldBe` 2
        genRandNum 1 2 10  `shouldBe` 1

    describe "readInn" $ do
      it "Checks if string is legal and returns raw location data(an index number of board) for board" $ do
        readInn "1 2" `shouldBe` (3,"")
        readInn "1 3 right"  `shouldBe`(6,"right")
      context "if too many numbers are inputted" $
        it "returns err code, and roll is not sent" $
          readInn "1 3 3 right"  `shouldBe` (11,"")
      context "if too high numbers are inputted" $
        it "returns err code, and roll is not sent" $
          readInn "9 9 right"  `shouldBe` (12,"")

    describe "readStr" $ do
      it "reads 1 and last string" $ do
        readStr "1 2" `shouldBe` ("1","2")

    describe "copErr" $ do
      it "returns fatal err on COP" $ do
        copErr `shouldBe` (False, '-')

    describe "readCOP" $ do
      it "reads string if player wants computer operated opponent" $ do
        readCOP "mongobongo" `shouldBe` copErr
      context "if no is inputted" $
        it "returns false for 1vCOP gameplay" $
          readCOP "no" `shouldBe` (False, '+')
      context "if legal yes is inputted" $
        it "returns true for 1vCOP gameplay and assigned role for COP" $
          readCOP "yes x" `shouldBe` (True,'O')
      context "if illegal yes is inputted" $
        it "returns fatal err on COP" $ do
          readCOP "yes" `shouldBe` copErr
          readCOP "yes mom" `shouldBe` copErr

    describe "copGetLegalSpace" $ do
      it "returns legal space on board for COP" $ do
        copGetLegalSpace createBoard (0,8,1) `shouldBe` 5
        copGetLegalSpace "XXXXXXXX " (0,8,1) `shouldBe` 8
      context "list has no spaces left" $
        it "returns unused err" $ do
          copGetLegalSpace "XXXXXXXXX" (0,8,1) `shouldBe` -1

    describe "copHandleRoll" $ do
      it "returns roll for COP" $ do
        copHandleRoll 1 `shouldBe` "right"
        copHandleRoll 6 `shouldBe` "left"
        copHandleRoll 2 `shouldBe` ""


unitTests :: Test
unitTests = TestList [
  TestLabel "countOcc" $ TestCase $
   assertEqual "counts occurrances of a char from string" 2 $
   countOcc 'J' "Jason, Jason Bourne" ,

  TestLabel "createBoard" $ TestCase $
   assertEqual "creates a new board" "         " createBoard,

  TestLabel "updateBoard" $ TestCase $
   assertEqual "updates board based on index data and char" "X        "
   $ updateBoard createBoard (0,'X'),

  TestLabel "printBoard" $ TestCase $
   assertEqual "converts raw board data to printable kawaii string" " \t   1  2  3 \n#\t1 [O][ ][X]\n#\t2 [ ][O][X]\n#\t3 [X][ ][ ]"
   $ printBoard "O X OXX  ",

  TestLabel "swapEarlyMarks" $ TestCase $
   assertEqual "swaps the first and third mark on board" "X O OXX  "
   $ swapEarlyMarks "O X OXX  ",

  TestLabel "roll" $ TestCase $
   assertEqual "swaps the first and third mark on board then rolls board to a direction" "X X O  XO"
   $ roll "O X OXX  " "right",

  TestLabel "checkDraw" $ TestCase $
   assertEqual "checks board for draw" True
   $checkDraw "OXOXOXXOX" 0 ,

  TestLabel "checkBoard" $ TestCase $
   assertEqual "checks board for win condition" (True,'X',"O | O|X |")
   $ checkBoard "O X OXX X" ,

  TestLabel "readUnit"  $ TestCase $
   assertEqual "reads string and checks for err " 13
   $  readUnit "" ,

  TestLabel "getPlayerTurn" $ TestCase $
   assertEqual "gets current turn and returns current- and next turn" ('X','O')
   $ getPlayerTurn 'X' ,

  TestLabel "gameLoopErrHandlr" $ TestCase $
   assertEqual "Gets raw index data for board, and checks for err" False
   $ gameLoopErrHandlr 1,

  TestLabel "convertInn" $ TestCase $
   assertEqual "Converts input to data nessescary for board. and checks for err" (1,2,"right")
   $ convertInn ["1","2", "right"],

  TestLabel "genRandNum" $ TestCase $
   assertEqual "gets a random num based on min max and seed" 2
   $ genRandNum 1 2 1,

  TestLabel "readInn" $ TestCase $
   assertEqual "Checks if string is legal and returns raw location data(an index number of board) for board" (6,"right")
   $ readInn "1 3 right",

  TestLabel "readStr"$ TestCase $
   assertEqual "reads 1 and last string" ("1","2")
   $  readStr "1 2",

  TestLabel "copErr" $ TestCase $
   assertEqual  "returns fatal err on COP" (False, '-') copErr,

  TestLabel "readCOP" $ TestCase $
   assertEqual "reads string if player wants computer operated opponent" (True,'O')
   $ readCOP "yes x",

  TestLabel "copGetLegalSpace"$ TestCase $
   assertEqual "returns legal space on board for COP" 8
   $copGetLegalSpace "XXXXXXXX " (0,8,1),

  TestLabel "copHandleRoll" $ TestCase $
   assertEqual"returns roll for COP" "right"
   $ copHandleRoll 1 
  ]



src/Lib.hs
module Lib
    (updateBoard,
    countOcc,
    newBoard,
    roll,
    printBoard,
    checkBoard,
    readInn,
    readUnit,
    getPlayerTurn) where

import Data.List.Split (splitOn)
import Data.Char(toLower)

countOcc :: Eq a => a -> [a] -> Int
countOcc want [] = 0
countOcc want list = sum $
 map(const 1) $
 filter (==want) list

newBoard::[Char]
newBoard = [
  ' ',' ',' ',
  ' ',' ',' ',
  ' ',' ',' '
  ]

updateBoard :: [Char] -> (Int, Char) -> [Char]
updateBoard [] _ = []
updateBoard (_:xs) (0,a) = a:xs
updateBoard (x:xs) (n,a) =
  if n < 0
    then x:xs
    else x: updateBoard xs (n-1,a)

printBoard::[Char]->String
printBoard inn = [
  ' ',' ',' ',' ',' ','1',' ',' ','2',' ',' ','3',' ','\n',
  '#',' ','1',' ','[',inn!!0,']','[',inn!!1,']','[',inn!!2,']','\n',
  '#',' ','2',' ','[',inn!!3,']','[',inn!!4,']','[',inn!!5,']','\n',
  '#',' ','3',' ','[',inn!!6,']','[',inn!!7,']','[',inn!!8,']','\n'
  ]


roll:: [Char] -> String ->  [Char]
roll  inn dir
  | dir == "left" =
    [
      inn!!0,inn!!5,inn!!8,
      inn!!1,inn!!4,inn!!7,
      inn!!2,inn!!3,inn!!6
      ]
  | dir == "right" =
    [
      inn!!6,inn!!3,inn!!2,
      inn!!7,inn!!4,inn!!1,
      inn!!8,inn!!5,inn!!0
      ]
  | otherwise = inn


checkBoard:: [Char] -> (Bool,Char,[Char])
checkBoard inn
  |   inn!!0 /= ' ' && inn!!0 == inn!!1 && inn!!0 == inn!!2 = (True,inn!!0,[
      '-','-','-',
      inn!!3,inn!!4,inn!!5,
      inn!!6,inn!!7,inn!!8
  ])
  |   inn!!3 /= ' ' && inn!!3 == inn!!4 && inn!!3 == inn!!5 = (True,inn!!3,[
      inn!!0,inn!!1,inn!!2,
      '-','-','-',
      inn!!6,inn!!7,inn!!8
  ])
  |   inn!!6 /= ' ' && inn!!6 == inn!!7 && inn!!6 == inn!!8 = (True,inn!!2,[
      inn!!0,inn!!1,inn!!2,
      inn!!3,inn!!4,inn!!5,
     '-','-','-'
  ])
  |   inn!!0 /= ' ' && inn!!0 == inn!!3 && inn!!0 == inn!!6 = (True,inn!!0,[
      '|',inn!!1,inn!!2,
      '|',inn!!4,inn!!5,
      '|',inn!!7,inn!!8
  ])
  |   inn!!1 /= ' ' && inn!!1 == inn!!4 && inn!!1 == inn!!7 = (True,inn!!1,[
      inn!!0,'|',inn!!2,
      inn!!3,'|',inn!!5,
      inn!!6,'|',inn!!8
  ])
  |   inn!!2 /= ' ' && inn!!2 == inn!!5 && inn!!2 == inn!!8 = (True,inn!!2,[
      inn!!0,inn!!1,'|',
      inn!!3,inn!!4,'|',
      inn!!6,inn!!7,'|'
  ])
  |   inn!!0 /= ' ' && inn!!0 == inn!!4 && inn!!0 == inn!!8 = (True,inn!!0,[
      '\\',inn!!1,inn!!2,
      inn!!3,'\\',inn!!5,
      inn!!6,inn!!7,'\\'
  ])
  |   inn!!2 /= ' ' && inn!!2 == inn!!4 && inn!!2 == inn!!6 = (True,inn!!2,[
      inn!!0,inn!!1,'/',
      inn!!3,'/',inn!!5,
      '/',inn!!7,inn!!8
  ])
  | otherwise = (False,' ',[])


type Unit = String

readUnit :: String -> Int
readUnit s = case reads s of               
    (n, ' ':unit):_ ->  n 
    (n, ""      ):_ ->  n  
    _ -> 13 -- Err msg

getPlayerTurn :: Char -> (Char, Char)
getPlayerTurn turn = do
  if turn == 'O' then
    ('O','X')
  else
    ('X','O')

readInn::String -> (Int,String)
readInn inn = do
  let (gArr,val1,val2,gLast) = (splitOn " " inn, readUnit $ gArr!!0::Int,readUnit $ gArr!!1::Int,map toLower (gArr!!(length gArr-1)))
  if val1 /= 13 || val2 /= 13 then do
    if length gArr == 2 then do
      if val2 == 0 || val1 == 0 || val1 > 3 || val2 > 3 then
        (12,"") else -- Err msg
          (3*(val2-1) + (val1-1),"")
    else if length gArr == 3 then do
      if val2 == 0 || val1 == 0 || val1 > 3 || val2 > 3 then
        (12,"") else -- Err msg
          if gLast == "right" then
            (3*(val2-1) + (val1-1),"right")
          else if gLast == "left" then
            (3*(val2-1) + (val1-1),"left")
          else
            (3*(val2-1) + (val1-1),"")
    else
      (11,"") -- Err msg
  else
    (13,"")  -- Err msg
  

app/Main.hs
module Main where

import Data.Time ( getCurrentTime, UTCTime(utctDayTime) )
import GameHandler (
    printBoard,
    readInn,
    updateBoard,
    roll,
    createBoard,
    checkBoard,
    gameLoopErrHandlr,
    getPlayerTurn,
    genRandNum,
    checkDraw
  )
import ComputerOperatedPlayer (
    readCOP,
    copHandleRoll,
    copGetLegalSpace,
    copErr
  )


gamePrompt::String->IO()
gamePrompt inn = do
    putStrLn "--------------------------------------------------------------------------------"
    putStrLn $ "\t"++ inn
    putStrLn "--------------------------------------------------------------------------------"


handleErr::Int->IO()
handleErr err = case err of
  11 ->
    gamePrompt "!!! Wrong amount of numbers, please input (x y) where the size is between 1-3"
  12 ->
    gamePrompt "!!! Wrong size of numbers, please input (x y) where the size is between 1-3"
  13 ->
    gamePrompt "!!! Wrong input, please input (x y) where x y is a number(1-3)"
  _ -> putStr ""

playerTurnHandler ::[Char]-> (Int,String)-> Char -> Char -> (Bool,Char) ->IO()
playerTurnHandler board (loc,rotation) pTurn wantMark cop = do
  if wantMark /= nextTurn then do -- checks if space is occupied
    putStrLn $ printBoard newBoard
    if fst cop then (do
      case rotation of
        "right" -> putStrLn ">>YOU -> Turning right..."
        "left" -> putStrLn  ">>YOU -> Turning left..."
        _-> putStr "") else (do
      case rotation of
        "right" -> putStrLn $ ">> ["++newMark:"] -> Turning right..."
        "left" -> putStrLn  $ ">> ["++newMark:"] -> Turning left..."
        _-> putStr "")
    gameLoop (roll newBoard rotation) nextTurn cop 0
  else do
    gamePrompt "!! Space is occupied!"
    gameLoop (roll board rotation) newMark cop 0
  where
    (newMark, nextTurn) = getPlayerTurn pTurn
    newBoard = updateBoard board (loc, newMark)

playerTurnPrompt ::[Char] -> Char -> (Bool,Char) -> IO String
playerTurnPrompt board turn cop = do
  if not $ fst cop then
   gamePrompt $ ">>> "++ turn : " - Turn"
  else
    gamePrompt $ ">>> Your("++ turn : ") Turn"
  putStrLn "\tIn order to make a move, write it as if it were coordinates on a map:\n \t\t x y , for example: 1 2\n"
  putStrLn $ printBoard board
  getLine

handlePlayer :: [Char]-> Char-> (Bool,Char) -> IO()
handlePlayer board turn cop = do
  inn <- playerTurnPrompt board turn cop
  let (tInn,rotation) = readInn inn
  if gameLoopErrHandlr tInn then do -- handles err
    handleErr tInn
    gameLoop board turn cop 0
  else do
    if (turn == 'X') || (turn == 'O') then (do -- handles player turn
      playerTurnHandler board (tInn,rotation) turn (board!!tInn) cop) else (do
      putStrLn "$!¤? FATAL ERROR <<<")



copTurnHandler:: [Char]-> Char-> (Bool,Char) -> Int->IO()
copTurnHandler board turn cop seed= do
  gamePrompt $ ">>> COP("++ turn : ") Turn"
  putStrLn "\tComputer Operated Player\n"
  putStrLn $ printBoard board
  putStrLn $ printBoard newBoard
  case turnHandler of
    "right" -> putStrLn ">>COP -> Turning right..."
    "left" -> putStrLn  ">>COP -> Turning left..."
    _-> putStr ""
  gameLoop (roll newBoard turnHandler ) nextTurn cop seed
  where
    (newMark, nextTurn) = getPlayerTurn $ snd cop
    (min, max) = (0::Int,8::Int)
    turnHandler = copHandleRoll seed
    loc =  copGetLegalSpace board (min,max,seed)
    newBoard = updateBoard board (loc, newMark)



handleCOP :: [Char]-> Char-> (Bool,Char)-> Int -> IO()
handleCOP board turn cop seed = do
  if (turn == 'X') || (turn == 'O') then (do -- handles Computer Operated Player turn
    copTurnHandler board turn cop seed) else (do
    putStrLn "$!¤? FATAL ERROR <<<")

playerAssignmentPrompt :: IO String
playerAssignmentPrompt = do
  gamePrompt "Do you want to play against a Computer Operated Player?\n\tInput either yes and the symbol you want, or no\n \t\t for example: YES X, yes o, no,NO"
  getLine

handleWinMsg:: Char -> (Bool,Char) -> IO ()
handleWinMsg winner cop = do
  if fst cop then
    if snd cop == winner then
      gamePrompt "!COP - The Computer Operated Player WON!"
    else
      gamePrompt "!YOU - You WON!"
  else
    gamePrompt $ "!Congrats [" ++ winner: "] for winning!"



gameLoop::[Char]-> Char-> (Bool,Char) -> Int ->IO()
gameLoop board turn cop seed = do
  if not winCondition then do  -- win condition checks recursively
    if not (checkDraw board 0) then do
      putStrLn $ "["++board++"]"
      if turn /= snd cop then do
        handlePlayer board turn cop
      else do
        handleCOP board turn cop seed
    else do
      putStrLn $ printBoard board
      gamePrompt "!DRAW!"
  else do
    putStrLn $ "["++board++"]"
    putStrLn $ printBoard winningBoard
    handleWinMsg winner cop
  where
    (winCondition,winner,winningBoard) = checkBoard board -- handles winning condition

startGameLoop :: Int -> IO()
startGameLoop seed = do
  gameType <- playerAssignmentPrompt
  let copData = readCOP gameType
  if copData /= copErr  then
    gameLoop newBoard 'X' copData seed
  else
    startGameLoop seed
  where
    newBoard = createBoard

main :: IO ()
main = do
  currTime<- getCurrentTime
  let seed = floor $ utctDayTime currTime ::Int
  startGameLoop seed
    

src/ComputerOperatedPlayer.hs
module ComputerOperatedPlayer(
    readStr,
    readCOP,
    copHandleRoll,
    copGetLegalSpace,
    copErr
  )where
import Data.List.Split ( splitOn )
import Data.Type.Equality (inner)
import Data.Char (toLower)
import GameHandler (
    genRandNum,
    createBoard,            
    checkDraw)

-- | readStr
-- >>> readStr "1 2" 
-- ("1","2")
readStr :: String -> (String,String)
readStr inn = do
  let list = splitOn " " inn
  (list!!0, list!!(length list-1))

-- | copErr
-- >>> copErr 
-- (False,'-')
copErr::(Bool,Char)
copErr = (False, '-')

-- | readCOP
-- >>> readCOP "mongobongo"  
-- (False,'-')
-- >>>readCOP "no" 
-- (False,'+')
-- >>>readCOP "yes x" 
-- (True,'O')
-- >>>readCOP "YES O" 
-- (True,'X')
-- >>>readCOP "yes" 
-- (False,'-')
-- >>>readCOP "yes mom" 
-- (False,'-')
readCOP::String -> (Bool,Char)
readCOP inn = do
  let ((t1,t2), first, second) = (readStr inn,map toLower t1, map toLower t2)
  case first of
    "yes" -> do
      if second == "x" then
        (True,'O')
      else if second == "o" then
        (True,'X')
      else
        copErr

    "no" -> (False,'+')
    _ -> copErr

-- | copGetLegalSpace
-- >>> copGetLegalSpace createBoard (0,8,1) 
-- 5
-- >>> copGetLegalSpace "XXXXXXXX " (0,8,1) 
-- 8
-- >>> copGetLegalSpace "XXXXXXXXX" (0,8,1) 
-- -1
copGetLegalSpace:: [Char]-> (Int, Int,Int) -> Int
copGetLegalSpace board (min,max,seed) = do
  if not (checkDraw board 0) then do
    let  rnd  = genRandNum min max seed
    if board!!rnd /= ' ' then  copGetLegalSpace board (min,max,seed +5)
    else rnd
  else -1

-- | copHandleRoll
-- >>> copHandleRoll 1 
-- "right"
-- >>> copHandleRoll 6 
-- "left"
-- >>> copHandleRoll 2 
-- ""
copHandleRoll ::Int -> String 
copHandleRoll seed = do
  case rnd of
    0 -> "left"
    1 -> "right"
    _->  ""
  where
    rnd  = genRandNum 0 3 seed





src/GameHandler.hs
module GameHandler(
      updateBoard,
      countOcc,
      createBoard,
      convertInn,
      roll,
      printBoard,
      swapEarlyMarks,
      checkBoard,
      readInn,
      readUnit,
      getPlayerTurn,
      gameLoopErrHandlr,
      genRandNum,
      checkDraw
    ) where

import Data.List.Split (splitOn)
import Data.Char(toLower)
import Data.Type.Equality (inner)
import System.Random (randomR, mkStdGen)

-- Pure game logic

-- | countOcc
--    counts occurances of an element from list
-- Examples:
--
-- >>> countOcc 'J' "Jason, Jason Bourne" 
-- 2
-- >>>  countOcc 'B' "BOB DYLAN IS OP"
-- 2
-- >>>  countOcc '!' ""
-- 0
countOcc :: Eq a => a -> [a] -> Int
countOcc want [] = 0
countOcc want list = sum $
 map(const 1) $
 filter (==want) list


-- | createBoard
--    sets a new board
-- Examples:
--
-- >>> createBoard 
-- "         " 
createBoard::[Char]
createBoard = [
  ' ',' ',' ',
  ' ',' ',' ',
  ' ',' ',' '
  ]


-- | updateBoard
--    Updates the board by replacing a value on the new board
--    (x:xs): the original board it takes inn(rawdata)
-- Examples:
--
-- >>> updateBoard createBoard (0,'X')
-- "X        "
-- >>> updateBoard "O X OXX  " (0,'X')
-- "X X OXX  "
updateBoard :: [Char] -> (Int, Char) -> [Char]
updateBoard [] _ = []
updateBoard (_:xs) (0,a) = a:xs
updateBoard (x:xs) (n,a) =
  if n < 0
    then x:xs
    else x: updateBoard xs (n-1,a)

-- | printBoard
--    converts raw board data to printable kawaii string
-- Examples:
--
-- >>> printBoard createBoard
-- " \t   1  2  3 \n#\t1 [ ][ ][ ]\n#\t2 [ ][ ][ ]\n#\t3 [ ][ ][ ]"
-- >>> printBoard "O X OXX  "
-- " \t   1  2  3 \n#\t1 [O][ ][X]\n#\t2 [ ][O][X]\n#\t3 [X][ ][ ]"
printBoard::[Char]->String
printBoard inn = [
  ' ','\t',' ',' ',' ','1',' ',' ','2',' ',' ','3',' ','\n',
  '#','\t','1',' ','[',inn!!0,']','[',inn!!1,']','[',inn!!2,']','\n',
  '#','\t','2',' ','[',inn!!3,']','[',inn!!4,']','[',inn!!5,']','\n',
  '#','\t','3',' ','[',inn!!6,']','[',inn!!7,']','[',inn!!8,']'
  ]


-- | swapEarlyMarks
--    "swaps the first and third mark on board
--    inn: the original board it takes inn(rawdata)
-- Examples:
--
-- >>> swapEarlyMarks createBoard 
-- "         "
-- >>> swapEarlyMarks "O X OXX  " 
-- "X O OXX  "
swapEarlyMarks::[Char] -> [Char]
swapEarlyMarks inn = [
  inn!!2,inn!!1,inn!!0,
  inn!!3,inn!!4,inn!!5,
  inn!!6,inn!!7,inn!!8
  ]

-- | roll
-- Rotates the board in a direction after swapping the 0th and 2nd mark:
-- inn: the original board it takes inn(rawdata)
-- dir: the direction of the rotation
--    
-- Examples:
--
-- >>> roll createBoard "left" 
-- "         "
-- >>> roll "O X OXX  " "right"
-- "X X O  XO"
-- >>> roll "O X OXX  " "left"
-- "OX  O X X"
roll:: [Char] -> String ->  [Char]
roll inn dir = do
  let new = swapEarlyMarks inn
  case dir of
    "left" ->   -- rotates board to the left
      [
        new!!2,new!!5,new!!8,
        new!!1,new!!4,new!!7,
        new!!0,new!!3,new!!6
        ]
    "right" ->  -- rotates board to the right
      [
        new!!6,new!!3,new!!0,
        new!!7,new!!4,new!!1,
        new!!8,new!!5,new!!2
        ]
    _ -> inn

-- | checkDraw
-- checks board for draw
-- >>> checkDraw createBoard 0 
-- False
-- >>> checkDraw "OXOXOXXOX" 0 
-- True
-- >>> checkDraw createBoard 0 
-- False
checkDraw:: [Char]->Int->Bool
checkDraw board index = index >= length board || (do (board!!index /= ' ') && checkDraw board (index+1))





-- | checkBoard
-- Checks board for win condition and returns:
-- if someone won  ::Bool
-- who won         ::Char
-- winning board   ::[Char] -- after drawing a line through the winning tiles (rawdata)
-- >>>  checkBoard createBoard
-- (False,' ',"")
-- >>> checkBoard "OXOXOXXOX"
-- (False,' ',"")
-- >>> checkBoard "O X OXX X"
-- (True,'X',"O | O|X |")
checkBoard:: [Char] -> (Bool,Char,[Char])
checkBoard inn
  |   inn!!0 /= ' ' && inn!!0 == inn!!1 && inn!!0 == inn!!2 = (True,inn!!0,[
      '-','-','-',
      inn!!3,inn!!4,inn!!5,
      inn!!6,inn!!7,inn!!8
  ])
  |   inn!!3 /= ' ' && inn!!3 == inn!!4 && inn!!3 == inn!!5 = (True,inn!!3,[
      inn!!0,inn!!1,inn!!2,
      '-','-','-',
      inn!!6,inn!!7,inn!!8
  ])
  |   inn!!6 /= ' ' && inn!!6 == inn!!7 && inn!!6 == inn!!8 = (True,inn!!6,[
      inn!!0,inn!!1,inn!!2,
      inn!!3,inn!!4,inn!!5,
     '-','-','-'
  ])
  |   inn!!0 /= ' ' && inn!!0 == inn!!3 && inn!!0 == inn!!6 = (True,inn!!0,[
      '|',inn!!1,inn!!2,
      '|',inn!!4,inn!!5,
      '|',inn!!7,inn!!8
  ])
  |   inn!!1 /= ' ' && inn!!1 == inn!!4 && inn!!1 == inn!!7 = (True,inn!!1,[
      inn!!0,'|',inn!!2,
      inn!!3,'|',inn!!5,
      inn!!6,'|',inn!!8
  ])
  |   inn!!2 /= ' ' && inn!!2 == inn!!5 && inn!!2 == inn!!8 = (True,inn!!2,[
      inn!!0,inn!!1,'|',
      inn!!3,inn!!4,'|',
      inn!!6,inn!!7,'|'
  ])
  |   inn!!0 /= ' ' && inn!!0 == inn!!4 && inn!!0 == inn!!8 = (True,inn!!0,[
      '\\',inn!!1,inn!!2,
      inn!!3,'\\',inn!!5,
      inn!!6,inn!!7,'\\'
  ])
  |   inn!!2 /= ' ' && inn!!2 == inn!!4 && inn!!2 == inn!!6 = (True,inn!!2,[
      inn!!0,inn!!1,'/',
      inn!!3,'/',inn!!5,
      '/',inn!!7,inn!!8
  ])
  | otherwise = (False,' ',[])

type Unit = String

-- | readUnit
-- reads string as Int, and returns err if anything other than int
-- >>> readUnit "1"  
-- 1
-- >>> readUnit "1 12311"
-- 1
-- >>> readUnit "1 12311a"  
-- 1
readUnit :: String -> Int
readUnit s = case reads s of
    (n, ' ':unit):_ ->  n
    (n, ""      ):_ ->  n
    _ -> 13 -- Err msg

-- |getPlayerTurn
-- Gets player turn and returns marks
-- turn: player turn
-- >>>  getPlayerTurn 'X'  
-- ('X','O')  
-- >>>  getPlayerTurn 'O'  
-- ('O','X') 
getPlayerTurn :: Char -> (Char, Char)
getPlayerTurn turn = do
  if turn == 'O' then -- if current player is O
    ('O','X')         -- returns current mark and next mark
  else                -- if current player is X
    ('X','O')         -- returns current mark and next mark

-- | gameLoopErrHandlr
-- Gets raw location data for board, and checks for err
-- err: raw location data for board
-- >>> gameLoopErrHandlr 11 
-- True  
-- >>>  gameLoopErrHandlr 12 
-- True  
-- >>>  gameLoopErrHandlr 13 
-- True 
-- >>>  gameLoopErrHandlr 1  
-- False
gameLoopErrHandlr :: Int -> Bool
gameLoopErrHandlr err = do
  (err == 11 || err == 12 || err == 13) && (do --different err codes
    True) -- returns true since err has been encountered

-- |convertInn
-- Converts input to data nessescary for board. and checks for err
-- inn: data from getLine as String 
-- >>> convertInn ["1","2"] 
-- (1,2,"2") 
-- >>> convertInn ["1","2", "right"]  
-- (1,2,"right")
-- >>> convertInn ["a","a"]  
-- (13,13,"a")
convertInn :: [String] -> (Int,Int,String)
convertInn inn = do
  let (
        first,  -- gets first value of string array as Int 
        second, -- gets second value of string array as Int 
        last    -- gets last value of string array as String 
        ) = (
          readUnit $ inn!!0::Int,
          readUnit $ inn!!1::Int,
          map toLower (inn!!(length inn-1))
          )
  (first,second,last)

-- | genRandNum
-- >>>  genRandNum 1 2 1  
-- 2  
-- >>>  genRandNum 1 2 10  
-- 1   
genRandNum:: Int -> Int -> Int -> Int
genRandNum min max seed = retNum
  where 
    (retNum,_) = randomR (min,max) $ mkStdGen seed

-- | readInn
-- Checks if string is legal and returns raw location data(an index number of board) for board
-- >>>  readInn "1 2" 
-- (3,"")  
-- >>>  readInn "1 3 right"  
-- (6,"right")
-- >>>  readInn "1 3 3 right"  
-- (11,"")
-- >>>  readInn "9 9 right"  
-- (12,"")
readInn::String -> (Int,String)
readInn inn = do
  let list = splitOn " " inn  --                        // Splits String into list of strings 
  let (
        val1, -- = readUnit $ list!!0::Int,             // Val1, first value of list is read as an int
        val2, -- = readUnit $ list!!1::Int,             // Val2, second value is read as an int
        gLast -- = map toLower (list!!(length list-1))  // gLast converts the last value to lowercase in order to get rotation 
        ) = convertInn list

  if val1 /= 13 || val2 /= 13 then do -- if a specific err hasnt been encountered
    if length list == 2 then do  -- if no
      if val2 == 0 || val1 == 0 || val1 > 3 || val2 > 3 then -- checks if numbers are legal
        (12,"") else -- Err msg
          (3*(val2-1) + (val1-1),"")        -- converts and returns coordinates to location in list
    else if length list == 3 then do
      if val2 == 0 || val1 == 0 || val1 > 3 || val2 > 3 then
        (12,"") else -- Err msg
          if gLast == "right" then          -- checks inputted rotation
            (3*(val2-1) + (val1-1),"right")  -- converts and returns coordinates to location in list and rotation
          else if gLast == "left" then      -- checks inputted rotation
            (3*(val2-1) + (val1-1),"left")   -- converts and returns coordinates to location in list and rotation
          else
            (3*(val2-1) + (val1-1),"")       -- converts and returns coordinates to location in list
    else
      (11,"")        -- Err msg
  else
      (13,"")        -- Err msg
  